e2e/internal/kubeclient/deploy.go

```diff
@@ -370,6 +405,80 @@ loop:
 	}
 }
 
+// WaitForEvent watches the EventList as long as the Event corresponding to the WaitCondition occurred after calling the function.
+func (c *Kubeclient) WaitForEvent(ctx context.Context, condition WaitCondition, resource ResourceWaiter, namespace, name string) error {
+	// The loop logic of this function is follows the pattern of waitFor(). Because we need to watch the EventList which should not implement the ResourceWaiter interface,
+	// we can not fall back to the resource watcher getter functions. Thus this function reimplements WaitFor() specifically designed to wait for specific ocuring events.
+
+	// StartingPoint is saved right here to ensure filtering only Events that happened after calling the function! The watcher keeps track of all events on the specific resource and
+	// as well calls the checking function when the EventList was changed regardless of the Event of our interest. Therefore avoidance of processing past Events has to be ensured, when
+	// checking the Event occurrence.
+	startingPoint := time.Now()
```

Comment:
Is this actually important? We're already watching a specific pod, and changing the policy annotation will schedule a new pod, so I'd expect all events for this pod to qualify.

Reply:
I slightly misinterpreted the event.Object in the watcher pattern, but i got it know:)

Reply:
![image](https://github.com/user-attachments/assets/b147fcc6-5dda-494f-9080-c40a512242b4)

Reply:
> The timestamp logic was introduced because the same event is reused in multiple waitForEvent instances.

Why do we need multiple `WaitForEvent` instances?

Reply:
The idea of checking for the event occurrence multiple times, was just to get more evidence that the container really don't start using the default policy. Somehow it looks like the watcher also reacts to older events or more precisely includes the current eventList as watch.ADDED events when instantiating the watcher. The timestamp logic only ensures that these watch.ADDED events are not seen as the event occurrence, based on the timestamp comparison.

Reply:
Did you know you can limit the events with https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions `SendInitialEvents` field?

I'm not really convinced that we need to observe the same failure more than once. If the concern is that it could fail once for an unrelated reason, but then succeed later on, should we improve the error message on the agent side instead and check for it?

Reply:
Oh no i didn't know, that actually looks like it bypasses the problem and the introduced timestamp logic can be dropped. 
Yes that was the initial concern, but the test logic like this introduces a lot of timing overhead just to observe the event multiple times. So i totally agree with you, that their might be a better solution for ensurance. Do you mean a specific error message that indicates the success later on?

Reply:
For my understanding we would need some message on agent side which determines that the default policy is still used and nevertheless the container was started? But i have no clue where to look at and how to check the policy annotation changes

Reply:
This could also be a later addition, I think it's fine to check for the existing error message. The message is from https://github.com/kata-containers/kata-containers/blob/cdaaf708a18da8e5f7e2b9824fa3e43b524893a5/src/agent/src/policy.rs#L37, you would need to add a patch on our side to change it - see https://github.com/edgelesssys/contrast/blob/main/dev-docs/patches.md.

Reply:
> Did you know you can limit the events with https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions `SendInitialEvents` field?
> 
> I'm not really convinced that we need to observe the same failure more than once. If the concern is that it could fail once for an unrelated reason, but then succeed later on, should we improve the error message on the agent side instead and check for it?

Using the SendInitialEvents parameter in the request, always rejects the request. It seems like watching an object and using the SendInitialEvents param, can not be combined: https://github.com/kubernetes-client/c/issues/203 . If you don't mind, i would keep the timestamp logic, because it is working perfectly fine.